ui: Add Storybook to the project

koorosh opened the pull request:
Depends on: https://github.com/cockroachdb/yarn-vendored/pull/16

- .storybook directory contains custom Storybook and Webpack configurations
- Added Badge stories as a working example
- "antd/es/tooltip/style/css" styles has been moved to the App component
because they apply globally.
- stick to the exact Storybook@5.0.6 version because of incompatibility with
babel-polyfill versions in latest versions of storybook.

koorosh committed:
ui: Add Storybook to the project

- .storybook directory contains custom Storybook and Webpack configurations
- Added Badge stories as a working example
- "antd/es/tooltip/style/css" styles has been moved to the App component
because they apply globally.
- stick to the exact Storybook@5.0.6 version because of incompatibility with
babel-polyfill versions in latest versions of storybook.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44680)
<!-- Reviewable:end -->

koorosh committed:
ui: Add copyright headers

Release note: None

koorosh committed:
ui: Fix yarn-lock and conflicts after rebasing against master branch

Release note: None

dhartunian commented:
@koorosh thanks for setting this up! I'm using this as an example to help set up the storybook for the `observability-ui` package that we'll have running soon in the `ui` repo. I'm still thinking through whether both repositories will benefit from having a Storybook instance as I don't want to introduce extra unnecessary dependencies into a repository that we're trying to remove code from but this config seems lightweight so it might be useful to have around.

Either way, I'm embarrassed to admit I have no idea how to run this storybook instance. Can you provide some documentation or a yarn command? I did run into some trouble in my configuration with getting all the CSS to line up correctly so it would be good to compare.

koorosh commented:
> @koorosh thanks for setting this up! I'm using this as an example to help set up the storybook for the `observability-ui` package that we'll have running soon in the `ui` repo. I'm still thinking through whether both repositories will benefit from having a Storybook instance as I don't want to introduce extra unnecessary dependencies into a repository that we're trying to remove code from but this config seems lightweight so it might be useful to have around.
> 
> Either way, I'm embarrassed to admit I have no idea how to run this storybook instance. Can you provide some documentation or a yarn command? I did run into some trouble in my configuration with getting all the CSS to line up correctly so it would be good to compare.

@dhartunian , storybook can be ran with following command: 
```
yarn run start-storybook -p 9009
```
- This starts Storybook with configuration provided in `.storybook/config.tsx`
- Take into account that global styles has to be imported in `.storybook/config.tsx` so they'll be 
applied for every story.
- Another tricky place is webpack configuration provided by Storybook. It has predefined `module/rules` to load styles. I removed it by providing only rules from `webpack.app.js`:

`.storybook/webpack.config.js`
```
  module: {
      rules: [
        ...appConfig.module.rules,
      ]
    },
```

dhartunian commented on pkg/ui/src/app.tsx:
```diff
@@ -54,6 +54,7 @@ import { ConnectedDecommissionedNodeHistory } from "src/views/reports";
 
 import "nvd3/build/nv.d3.min.css";
 import "react-select/dist/react-select.css";
+import "antd/es/tooltip/style/css";
```
Why is this here?

koorosh replied:
@dhartunian 
1. Badges styles are not 100% ready to be honest, I picked this component as easiest one for demo purposes only. It has the same colors because not all styles are defined.
It is already done in master.

2. Component width in Firefox. It is really a problem - **nib** library (extension for stylus) does not add vendor prefixes for **width** attribute with `max-content` or `fit-content` values.
I will create separate issue on this because it doesn't relate to Storybook integration

3. "antd/es/tooltip/style/css" this is a global style on which we depend. Previously it was imported in Tooltip component but was applied globally. To make it explicit that it is global styles I put it in top level component.

koorosh replied:
3. (Updated) you're right, it's unnecessary to move styles import to App.tsx. Reverted this change back.

dhartunian reviewed (changes requested):
@koorosh can you take another look at the CSS here?

The badges all look the same in chrome for me:
<img width="577" alt="Screen Shot 2020-03-02 at 10 21 57 AM" src="https://user-images.githubusercontent.com/986307/75689771-c1c10880-5c6f-11ea-9fa6-5ee3aaf33875.png">
<img width="566" alt="Screen Shot 2020-03-02 at 10 21 54 AM" src="https://user-images.githubusercontent.com/986307/75689773-c2599f00-5c6f-11ea-9285-951a8757ee11.png">

And in firefox, the badges are full width
<img width="1055" alt="Screen Shot 2020-03-02 at 10 22 07 AM" src="https://user-images.githubusercontent.com/986307/75689802-cf768e00-5c6f-11ea-8aff-4afad7055517.png">

koorosh committed:
ui: Resolve conflicts after rebase
`window.setTimeout` change allows correctly suggest
type for `this.timeoutHandler`. Project build failed
without this fix.

Release note: None

koorosh committed:
ui: Add storybook script to package.json

Storybook is aimed to be run locally during
development so there is no needs to define it
in Makefile, instead of that it can be run as
NPM script.

Release note: None

dhartunian reviewed (approved):
Thanks for the final change. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 8 of 8 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, 2 of 2 files at r7, 3 of 3 files at r9, 1 of 1 files at r10, 2 of 2 files at r11, 3 of 3 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44680#-:-M4VdxiyCdUGWJiii20n:b-pyog4w)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian)



<!-- Sent from Reviewable.io -->

dhartunian commented:
bors r+
